Lib/test/test_typing.py

```diff
@@ -1483,14 +1483,18 @@ def test_cannot_instantiate(self):
             type(u)()
 
     def test_cannot_iterate(self):
-        with self.assertRaises(TypeError):
+        expected_message = "^'_SpecialForm' object is not iterable$"
```

Comment:
`_SpecialForm` is an implementation detail. I think it is enough to test for "is not iterable".

Reply:
You mean, whether we should use `assertRaisesRegex` or just `assertRaises` here?

I was thinking it might be better to use `assertRaisesRegex` to make it obvious to readers of the test exactly what behaviour this is supposed to produce - since, from a user's perspective, the error message is a key part of the behaviour here. In particular, I wanted this test to be honest about the fact that the error messages are a bit confusing - as you say, `_SpecialForm` is an implementation detail which most users won't know about. (Ideally maybe we'd also improve these error messages, but since the message seems to come from C, I'm guessing it would be hard. And we don't need to be perfectionist here; the fact that this is only being fixed now is evidence that this isn't a common thing for users to run into.)

On the other hand, one could argue that tests communicate "The behaviour we want to make _sure_ is the case" - and since we're not committed to these particular error messages, the tests could erroneously communicate the wrong intent.

I don't have strong feelings, though - happy to defer to whatever you think is best.

Reply:
I agree with Serhiy -- I think we should use `assertRaisesRegex`, but there's no need to use the exact name of the class in the assertion. We can just do this:

```python
class TypeIterationTests(BaseTestCase):
    _UNITERABLE_OBJECTS = (
        Any,
        Union,
        Union[str, int],
        Union[str, T],
        List,
        Tuple,
        Callable,
        Callable[..., T],
        Callable[[T], str],
        Annotated,
        Annotated[T, '']
    )

    def test_cannot_iterate(self):
        for test_type in self._UNITERABLE_OBJECTS:
            err_message = 'object is not iterable'
            with self.subTest(type=test_type):
                with self.assertRaisesRegex(TypeError, err_message):
                    iter(test_type)
                with self.assertRaisesRegex(TypeError, err_message):
                    list(test_type)
                with self.assertRaisesRegex(TypeError, err_message):
                    for _ in test_type:
                        pass

    def test_is_not_instance_of_iterable(self):
        for type_to_test in self._UNITERABLE_OBJECTS:
            self.assertNotIsInstance(type_to_test, collections.abc.Iterable)
```

That way the test won't have to be needlessly refactored if we refactor some of the internal implementation details of the `typing` module at some later date. It also makes the test slightly simpler, as a bonus :)

Reply:
`assertRaisesRegex(TypeError, "is not iterable")`

These are the most important and the most stable parts of the error: 1) that it is a TypeError; 2) that the error message is specialized for iteration, it is not generic "'NoneType' object is not callable".

Reply:
Ohhh, sorry, gotcha - I'd misunderstood. Fixed.